### ISSUE TITLE
Compile fails with -Werror=unused-function

### ISSUE THREAD

**Issue by dwmw2** [svaarala/duktape#916]
I imported duktape into PacRunner to handle proxy PAC scripts.

The build fails because it uses -Werror=unused-function:
```
duktape/duktape.c:9796:19: error: ‘duk_err_unsupported_defmsg’ defined but not used [-Werror=unused-function]
DUK_INTERNAL void duk_err_unsupported_defmsg(duk_hthread *thr, const char *filename, duk_int_t linenumber) {
```

For the GCC/DUK_SINGLE_FILE build, this can be avoided by adding `__attribute__((unused))` to the definition of `DUK_INTERNAL`, thus:
`#define DUK_INTERNAL static __attribute__((unused))`

**Comment by svaarala** [comment]
Duktape doesn't attempt to guarantee a warning-free compilation - it's very difficult when supporting a lot of compilers (sometimes the needed fixes conflict between two compilers, for example). But I'll of course try to fix any warnings that pop up in practice.

For this particular case, the issue has been fixed in master. It's present in 1.5.0 and e.g. Clang warns about it (gcc 4.8.4 does not). I'll make a note to fix this in 1.5 maintenance.

**Comment by dwmw2** [comment]
Understood. And making functions *actually* disappear just because there don't happen to be any callers would be really hard, potentially involving a horrid mess of #ifdefs. 

But adding __attribute__((unused)) to DUK_INTERNAL seems like a simple enough and non-intrusive solution.

**Comment by svaarala** [comment]
Hmm, GCC documentation says:

> This attribute, attached to a variable, means that the variable is meant to be possibly unused. GCC will not produce a warning for this variable. 

So maybe that wouldn't be too much of a stretch. Same for Clang:

> When passing the -Wunused flag to Clang, entities that are unused by the program may be diagnosed. The `[[maybe_unused]]` (or `__attribute__((unused))`) attribute can be used to silence such diagnostics when the entity cannot be removed. For instance, a local variable may exist solely for use in an assert() statement, which makes the local variable unused when NDEBUG is defined.

That matches the Duktape portability issue - and there are in fact many variables which are defined only when assertions are enabled for example (not just functions).

I'll see if this could be worked in, for functions and also for local variables which are config dependent. It's indeed a bit painful to manage the #ifdef conditions manually :)

I'll also need a solution for MSVC to be able to actually omit those #ifdefs though.

**Comment by svaarala** [comment]
This should now be fixed in master, v1-maintenance, and v1.5-maintenance. There's no MSVC equivalent yet however so there's still a need to try to keep unused function #ifdefs in sync with their call sites.
